source/Font.cpp

```diff
@@ -12,91 +12,169 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
-		// Aspect ratio of rendered glyph (unity by default).
-		"uniform float aspect = 1.f;\n"
+	// Determines the number of bytes used by the unicode code point in utf8.
+	int CodePointBytes(const char *str)
+	{
+		// end - 00000000
+		if(!str || !*str)
+			return 0;
 		
-		// Inputs from the VBO.
-		"in vec2 vert;\n"
-		"in vec2 corner;\n"
+		// 1 byte - 0xxxxxxx
+		if((*str & 0x80) == 0)
+			return 1;
 		
-		// Output to the fragment shader.
-		"out vec2 texCoord;\n"
+		// invalid - 10?????? or 11?????? invalid
+		if((*str & 0x40) == 0 || (*(str + 1) & 0xc0) != 0x80)
+			return -1;
 		
-		// Pick the proper glyph out of the texture.
-		"void main() {\n"
-		"  texCoord = vec2((glyph + corner.x) / 98.f, corner.y);\n"
-		"  gl_Position = vec4((aspect * vert.x + position.x) * scale.x, (vert.y + position.y) * scale.y, 0, 1);\n"
-		"}\n";
-	
-	const char *fragmentCode =
-		// The user must supply a texture and a color (white by default).
-		"uniform sampler2D tex;\n"
-		"uniform vec4 color = vec4(1, 1, 1, 1);\n"
+		// 2 bytes - 110xxxxx 10xxxxxx
+		if((*str & 0x20) == 0)
+			return 2;
 		
-		// This comes from the vertex shader.
-		"in vec2 texCoord;\n"
+		// invalid - 111????? 10?????? invalid
+		if((*(str + 2) & 0xc0) != 0x80)
+			return -1;
 		
-		// Output color.
-		"out vec4 finalColor;\n"
+		// 3 bytes - 1110xxxx 10xxxxxx 10xxxxxx
+		if((*str & 0x10) == 0)
+			return 3;
 		
-		// Multiply the texture by the user-specified color (including alpha).
-		"void main() {\n"
-		"  finalColor = texture(tex, texCoord).a * color;\n"
-		"}\n";
-	
-	const int KERN = 2;
+		// invalid - 1111???? 10?????? 10?????? invalid
+		if((*(str + 3) & 0xc0) != 0x80)
+			return -1;
+		
+		// 4 bytes - 11110xxx 10xxxxxx 10xxxxxx 10xxxxxx
+		if((*str & 0x8) == 0)
+			return 4;
+		
+		// not unicode - 11111??? 10?????? 10?????? 10??????
+		return -1;
+	}
 }
 
 
 
 Font::Font()
-	: texture(0), vao(0), vbo(0), colorI(0), scaleI(0), glyphI(0), aspectI(0),
-	  positionI(0), height(0), space(0), screenWidth(0), screenHeight(0)
+	: size(0)
 {
 }
 
 
 
-Font::Font(const string &imagePath)
-	: Font()
+bool Font::Load(const DataNode &node)
 {
-	Load(imagePath);
+	// Ignore if already loaded.
+	if(!sources.empty())
+		return false;
+	
+	if(node.Size() < 1 || node.Token(0) != "font")
+	{
+		node.PrintTrace("was expecting font");
+		return false;
+	}
+	
+	// Get size.
+	size = 0;
+	for(const DataNode &child : node)
+	{
+		string key = (child.Size() >= 0 ? child.Token(0) : "");
+		if(key != "size")
+			continue;
+		
+		if(size > 0)
+		{
+			node.PrintTrace("too many sizes");
+			return false;
+		}
+		size = round(child.Value(1));
+		if(size <= 0)
+		{
+			child.PrintTrace("invalid font.size");
+			return false;
+		}
+	}
+	if(size <= 0)
+	{
+		node.PrintTrace("missing size");
+		return false;
+	}
+	
+	// Get glyph source.
+	sources.clear();
+	for(const DataNode &child : node)
+	{
+		if(child.Size() <= 0 || (child.Token(0) != "atlas" && child.Token(0) != "freetype"))
```

Comment:
```
static const set<string> allowed = {"atlas", "freetype"};
if(!allowed.count(child.Token(0))
    continue;
```

Reply:
Just for 2? I rather not.

Reply:
Mostly I just hate seeing repeated `child.Token(0)...child.Token(0)...` etc. For 2, yes, using a `set.count()` check is perhaps a bit overzealous. It's certainly expressive :+1: